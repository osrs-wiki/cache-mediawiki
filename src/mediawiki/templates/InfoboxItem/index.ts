export { default as InfoboxItem } from "./InfoboxItem";
