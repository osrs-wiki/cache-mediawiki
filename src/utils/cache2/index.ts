export * from "./LazyPromise";
export * from "./types";
