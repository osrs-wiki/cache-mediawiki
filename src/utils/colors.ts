export const jagexHSLtoHex = (value: number) => {
  const hue = (value >> 10) & 0x3f;
  const sat = (value >> 7) & 0x07;
  const lum = value & 0x7f;

  console.log(`H: ${hue}, S: ${sat}, L: ${lum}`);
  return hslToHex(hue, sat, lum);
};

export const hslToHex = (h: number, s: number, l: number) => {
  l /= 100;
  const a = (s * Math.min(l, 1 - l)) / 100;
  const f = (n: number) => {
    const k = (n + h / 30) % 12;
    const color = l - a * Math.max(Math.min(k - 3, 9 - k, 1), -1);
    return Math.round(255 * color)
      .toString(16)
      .padStart(2, "0"); // convert to Hex and prefix "0" if needed
  };
  return `#${f(0)}${f(8)}${f(4)}`;
};
