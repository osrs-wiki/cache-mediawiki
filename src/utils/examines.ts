export const getExamines = async (
  name: string,
  version = "master"
): Promise<{ [key: string]: string }> => {
  const examines: { [key: string]: string } = {};
  const req = await fetch(
    `https://raw.githubusercontent.com/Joshua-F/osrs-examines/${version}/${name}.csv`
  );
  if (!req.ok) {
    return;
  }
  const fetchedExamines = await req.text();
  fetchedExamines.split("\n").forEach((line) => {
    const id = line.split(",")[0];
    const examine = line.substring(line.indexOf(",") + 1).replaceAll('"', "");
    if (id) {
      examines[id] = examine;
    }
  });
  return examines;
};
