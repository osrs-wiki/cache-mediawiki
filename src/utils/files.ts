export const formatFileName = (name: string) => {
  return name
    .replaceAll(":", "-")
    .replaceAll("?", "")
    .replaceAll(/<[^>]*>/g, "");
};
