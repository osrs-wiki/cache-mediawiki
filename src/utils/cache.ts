import { DiskCacheProvider, FlatCacheProvider } from "@abextm/cache2";
import { readFile } from "fs/promises";

export type CacheSource = "github" | "local";
export type CacheFileType = "flat" | "disk";

export const getCacheProviderLocal = async (
  version: string,
  type: CacheFileType
) => {
  if (type === "disk") {
    return getCacheProviderDiskLocal(version);
  } else if (type === "flat") {
    return getCacheProviderFlatLocal(version);
  }
};

export const getCacheProviderDiskLocal = async (version: string) => {
  return new DiskCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/diskcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderFlatLocal = async (version: string) => {
  return new FlatCacheProvider({
    async getFile(name) {
      const ab = (await readFile(`./data/flatcache/${version}/${name}`)).buffer;
      return new Uint8Array(ab);
    },
  });
};

export const getCacheProviderGithub = async (version = "master") => {
  return new FlatCacheProvider({
    async getFile(name) {
      const req = await fetch(
        `https://raw.githubusercontent.com/abextm/osrs-cache/${version}/${name}`
      );
      if (!req.ok) {
        return;
      }
      const ab = await req.arrayBuffer();
      return new Uint8Array(ab);
    },
  });
};
