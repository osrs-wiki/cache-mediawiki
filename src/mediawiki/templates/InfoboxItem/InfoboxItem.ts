import {
  InfoboxTemplate,
  MediaWikiDate,
  MediaWikiFile,
} from "@osrs-wiki/mediawiki-builder";
import type { InfoboxItem as InfoboxItemTemplate } from "@osrs-wiki/mediawiki-builder";

import { getInventoryActions } from "./InfoboxItem.utils";
import Context from "../../../context";
import { Item } from "../../../utils/cache2";

const InfoboxItem = (item: Item) => {
  return new InfoboxTemplate<InfoboxItemTemplate>("item", {
    name: item.name as string,
    image: new MediaWikiFile(`${item.name}.png`),
    release: Context.updateDate
      ? new MediaWikiDate(new Date(Context.updateDate))
      : undefined,
    update: Context.update,
    members: item.isMembers as boolean,
    quest: "No",
    exchange: item.isGrandExchangable,
    tradeable: item.isGrandExchangable,
    bankable: item.placeholderLinkedItem > 0,
    placeholder: item.placeholderLinkedItem > 1 ? true : undefined,
    equipable: item.wearpos1 > 0,
    stackable: item.isStackable,
    noteable: item.noteLinkedItem > 0,
    options: getInventoryActions(item),
    examine: item.examine,
    value: item.price,
    weight: (item.weight / 1000).toFixed(3),
    id: item.id.toString(),
  });
};

export default InfoboxItem;
