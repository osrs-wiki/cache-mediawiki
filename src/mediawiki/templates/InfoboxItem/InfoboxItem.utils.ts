import { MediaWikiFile, MediaWikiHTML } from "@osrs-wiki/mediawiki-builder";

import { Item } from "@/utils/cache2";

/**
 * Get the inventory actions for a given item.
 * @param item The item to get inventory actions for.
 * @returns An array of inventory actions, possibly with sub-operations included.
 */
export const getInventoryActions = (item: Item) => {
  return item.inventoryActions?.map((action, index) => {
    const subops = item.subops?.[index];
    return subops?.length > 0 ? `${action} (${subops.join(", ")})` : action;
  });
};

/**
 * Get the item infobox image for a given item, with support for stack variants.
 * If the item has no stack variants, it returns a single image.
 * If it has stack variants, it returns an array of images, inserting a break
 * after the first half if there are 5 or more variants.
 * @param item The item to get the infobox image for.
 * @returns The infobox image for the item.
 */
export const getItemInfoboxImage = (item: Item) => {
  const filteredVariants = item.stackVariantItems.filter((id) => id !== 0);
  if (filteredVariants.length === 0) {
    return new MediaWikiFile(`${item.name}.png`);
  } else {
    const variantFiles = filteredVariants.map(
      (_variantId, index) =>
        new MediaWikiFile(
          `${item.name} ${item.stackVariantQuantities[index]}.png`
        )
    );
    if (variantFiles.length >= 5) {
      const half = Math.floor(variantFiles.length / 2);
      return [
        ...variantFiles.slice(0, half),
        new MediaWikiHTML("br"),
        ...variantFiles.slice(half),
      ];
    } else {
      return variantFiles;
    }
  }
};
