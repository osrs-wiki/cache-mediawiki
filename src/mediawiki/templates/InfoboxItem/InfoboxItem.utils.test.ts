import { getInventoryActions } from "./InfoboxItem.utils";

describe("InfoboxItem utils", () => {
  test("getInventoryActions - no sub ops", () => {
    expect(
      getInventoryActions({
        inventoryActions: ["Test Action"],
        subops: [],
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any)
    ).toEqual(["Test Action"]);
  });

  test("getInventoryActions - with sub ops", () => {
    expect(
      getInventoryActions({
        inventoryActions: ["Test Action 1", "Test Action 2", "Test Action 3"],
        subops: [["Sub Action 1"], ["Sub Action 2.1", "Sub Action 2.2"], []],
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any)
    ).toEqual([
      "Test Action 1 (Sub Action 1)",
      "Test Action 2 (Sub Action 2.1, Sub Action 2.2)",
      "Test Action 3",
    ]);
  });
});
