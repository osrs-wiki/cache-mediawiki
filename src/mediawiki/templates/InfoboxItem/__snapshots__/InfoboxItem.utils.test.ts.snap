// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoboxItem utils getItemInfoboxImage inserts MediaWikiBreak correctly for odd number of variants 1`] = `
Array [
  MediaWikiFile {
    "children": undefined,
    "fileName": "Odd Item 5.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Odd Item 15.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Odd Item 25.png",
    "options": undefined,
  },
  MediaWikiHTML {
    "attributes": undefined,
    "children": undefined,
    "options": undefined,
    "tag": "br",
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Odd Item 35.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Odd Item 45.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Odd Item 55.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Odd Item 65.png",
    "options": undefined,
  },
]
`;

exports[`InfoboxItem utils getItemInfoboxImage returns MediaWikiFile when all stack variants are 0 1`] = `
MediaWikiFile {
  "children": undefined,
  "fileName": "Test Item.png",
  "options": undefined,
}
`;

exports[`InfoboxItem utils getItemInfoboxImage returns MediaWikiFile when no stack variants 1`] = `
MediaWikiFile {
  "children": undefined,
  "fileName": "Test Item.png",
  "options": undefined,
}
`;

exports[`InfoboxItem utils getItemInfoboxImage returns array with MediaWikiBreak inserted halfway 1`] = `
Array [
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 10.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 20.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 30.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 40.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 50.png",
    "options": undefined,
  },
  MediaWikiHTML {
    "attributes": undefined,
    "children": undefined,
    "options": undefined,
    "tag": "br",
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 60.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 70.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 80.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 90.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 100.png",
    "options": undefined,
  },
]
`;

exports[`InfoboxItem utils getItemInfoboxImage returns no break when less than 5 stack variants 1`] = `
Array [
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 10.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 20.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 30.png",
    "options": undefined,
  },
  MediaWikiFile {
    "children": undefined,
    "fileName": "Test Item 40.png",
    "options": undefined,
  },
]
`;
